Adding a version serialization type to ConvPackedParam

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42941 [DO NOT MERGE] Trying to redefine the __setstate__ using def_unboxed
* **#40003 Adding a version serialization type to ConvPackedParam**

Differential Revision: [D22158980](https://our.internmc.facebook.com/intern/diff/D22158980)

Commit:
Adding a version serialization type to ConvPackedParam

[ghstack-poisoned]

Commit:
Update on "Adding a version serialization type to ConvPackedParam"




[ghstack-poisoned]

Commit:
Update on "Adding a version serialization type to ConvPackedParam"




[ghstack-poisoned]

Commit:
Update on "Adding a version serialization type to ConvPackedParam"




[ghstack-poisoned]

Commit:
Update on "Adding a version serialization type to ConvPackedParam"




[ghstack-poisoned]

Commit:
Update on "Adding a version serialization type to ConvPackedParam"




[ghstack-poisoned]

Commit:
Update on "Adding a version serialization type to ConvPackedParam"




[ghstack-poisoned]

Commit:
Update on "Adding a version serialization type to ConvPackedParam"




[ghstack-poisoned]

Commit:
Update on "Adding a version serialization type to ConvPackedParam"




[ghstack-poisoned]

Commit:
Update on "Adding a version serialization type to ConvPackedParam"




[ghstack-poisoned]

Commit:
Update on "Adding a version serialization type to ConvPackedParam"




[ghstack-poisoned]

Commit:
Update on "Adding a version serialization type to ConvPackedParam"




[ghstack-poisoned]

Commit:
Update on "Adding a version serialization type to ConvPackedParam"




[ghstack-poisoned]

Commit:
Update on "Adding a version serialization type to ConvPackedParam"


Differential Revision: [D22158980](https://our.internmc.facebook.com/intern/diff/D22158980)

[ghstack-poisoned]

Commit:
Update on "Adding a version serialization type to ConvPackedParam"


Differential Revision: [D22158980](https://our.internmc.facebook.com/intern/diff/D22158980)

[ghstack-poisoned]

Commit:
Update on "Adding a version serialization type to ConvPackedParam"


Differential Revision: [D22158980](https://our.internmc.facebook.com/intern/diff/D22158980)

[ghstack-poisoned]

Commit:
Update on "Adding a version serialization type to ConvPackedParam"


Differential Revision: [D22158980](https://our.internmc.facebook.com/intern/diff/D22158980)

[ghstack-poisoned]

Commit:
Update on "Adding a version serialization type to ConvPackedParam"


Differential Revision: [D22158980](https://our.internmc.facebook.com/intern/diff/D22158980)

[ghstack-poisoned]

Commit:
Update on "Adding a version serialization type to ConvPackedParam"


Differential Revision: [D22158980](https://our.internmc.facebook.com/intern/diff/D22158980)

[ghstack-poisoned]

Commit:
Update on "Adding a version serialization type to ConvPackedParam"


Differential Revision: [D22158980](https://our.internmc.facebook.com/intern/diff/D22158980)

[ghstack-poisoned]

Commit:
Update on "Adding a version serialization type to ConvPackedParam"


Differential Revision: [D22158980](https://our.internmc.facebook.com/intern/diff/D22158980)

[ghstack-poisoned]

Commit:
Update on "Adding a version serialization type to ConvPackedParam"


Differential Revision: [D22158980](https://our.internmc.facebook.com/intern/diff/D22158980)

[ghstack-poisoned]

Commit:
Update on "Adding a version serialization type to ConvPackedParam"


Differential Revision: [D22158980](https://our.internmc.facebook.com/intern/diff/D22158980)

[ghstack-poisoned]

vkuzo commented:
I had a couple of thoughts here:
* would be great to have a dedicated 'version' field if we are changing the structure
* is it a requirement to use tuple, or can we use a dictionary insead?
* what are the pros/cons of adding dedicated fields (as opposed to packing multiple things into a single field)

looking forward to the doc / discussion!

Commit:
Update on "Adding a version serialization type to ConvPackedParam"


Differential Revision: [D22158980](https://our.internmc.facebook.com/intern/diff/D22158980)

[ghstack-poisoned]

Commit:
Update on "Adding a version serialization type to ConvPackedParam"


Differential Revision: [D22158980](https://our.internmc.facebook.com/intern/diff/D22158980)

[ghstack-poisoned]

vkuzo reviewed (commented):
just to confirm, we are actually changing the serialization format here, right?  In that case, would it be possible to add a test which loads the previous format as a part of this PR, to ensure BC?

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/fbgemm_utils.cpp:
```diff
@@ -212,67 +213,106 @@ Tensor ConvertToChannelsLast3dTensor(const Tensor& src) {
-          torch::List<at::Tensor> padding;
-          torch::List<at::Tensor> dilation;
-          at::Tensor groups;
+          std::vector<int64_t> scalar_elements;
```
what is the format we agreed upon in the end? Looks like MS is not going to support int64_t in ONNX, so we might need to keep Tensor for scalar elements

jerryzh168 replied:
Given that we might want to pack all the parameters into one entry to keep it compact. maybe we can pack stride/padding/dilation/transpose into one Tensor.

z-a-f replied:
I a mwriting a new serialization with more "scalable" and "reusable" serialization type. Will send an update within an hour

z-a-f replied:
So I see that RNN uses vector of double/int -- can we use that? Reference: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/RNN.cpp#L59-L64

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/fbgemm_utils.cpp:
```diff
@@ -212,67 +213,106 @@ Tensor ConvertToChannelsLast3dTensor(const Tensor& src) {
-          torch::List<at::Tensor> padding;
-          torch::List<at::Tensor> dilation;
-          at::Tensor groups;
+          std::vector<int64_t> scalar_elements;
```
actually does this work currently? i think this does not work with onnx

jerryzh168 replied:
if this works we can put everything in std::vector<int64_t>

Commit:
Update on "Adding a version serialization type to ConvPackedParam"


Differential Revision: [D22158980](https://our.internmc.facebook.com/intern/diff/D22158980)

[ghstack-poisoned]

jerryzh168 reviewed (approved):
Looks good, if this works.

vkuzo commented:
can we add a test in this PR please?

z-a-f commented:
> can we add a test in this PR please?

there are tests for the serialization -- backwards_compatibilitty tests take care of them

Commit:
Update on "Adding a version serialization type to ConvPackedParam"


Differential Revision: [D22158980](https://our.internmc.facebook.com/intern/diff/D22158980)

[ghstack-poisoned]

Commit:
Update on "Adding a version serialization type to ConvPackedParam"


Differential Revision: [D22158980](https://our.internmc.facebook.com/intern/diff/D22158980)

[ghstack-poisoned]

Commit:
Update on "Adding a version serialization type to ConvPackedParam"


Differential Revision: [D22158980](https://our.internmc.facebook.com/intern/diff/D22158980)

[ghstack-poisoned]

Commit:
Update on "Adding a version serialization type to ConvPackedParam"


Differential Revision: [D22158980](https://our.internmc.facebook.com/intern/diff/D22158980)

[ghstack-poisoned]

vkuzo commented on aten/src/ATen/native/quantized/cpu/serialization_versions.h:
```diff
@@ -0,0 +1,164 @@
+ *  5. dilation x kSpatialDim
+ *  6. groups
+ *
+ * Version 2+
```
nit: Version 2? versions higher than 2 should have a separate entry

vkuzo replied:
can we also add explicit types to the comments for all fields in v1 and v2?

vkuzo commented on aten/src/ATen/native/quantized/cpu/serialization_versions.h:
```diff
@@ -0,0 +1,164 @@
+
+template <uint32_t kSpatialDim>
+c10::intrusive_ptr<ConvPackedParamsBase<kSpatialDim>> deserialize_conv(
+    SerializationType state) {
```
might be a n00b question.  This looks like the logic to deserialize the v2 format.  Is there logic somewhere to deserialize the v1 format?

vkuzo commented on aten/src/ATen/native/quantized/cpu/serialization_versions.h:
```diff
@@ -0,0 +1,164 @@
+  std::vector<int64_t> longs;
+
+  std::tie(name, version, non_optional, optional, doubles, longs) = state;
+
```
should we enforce that name == "conv" here?

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/serialization_versions.h:
```diff
@@ -0,0 +1,164 @@
+#include <tuple>
+
+using SerializationType = std::tuple<
+  std::string /*name of the op that is serialized*/,
```
why do we serialize this?

vkuzo replied:
I assumed this would be if `SerializationType` is used for other ops in the future

jerryzh168 replied:
when this is used for other ops, why do we need to write down the name here?(just went through similar process with @supriyar 's PR actually). There are two questions to ask here:
(1). is it necessary to distinguish between different ops? if not, we can just remove it
(2). if so, can we put this as a template argument of the type? e.g. ConvPackedParams<1> ConvPackedParams<2> for conv1d and conv2d
(3). even we need to distinguish the op name and can't put it in template, I would consider serializing integer(enum) values better than serializing strings since this is replicated for all instances.

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/serialization_versions.h:
```diff
@@ -0,0 +1,164 @@
+ *    1. unpacked weight
+ *  4. list of all optional tensors (vector<optional<Tensor>>)
+ *    1. bias
+ *  5. list of all doubles
```
why do we have this now if this is empty?

z-a-f replied:
Because we have to have the same number of fields going into the future (forward compatibility). Otherwise, we will end up with the same problem if we decide to add doubles to the tuple

vkuzo replied:
>  the same number of fields

this seems pretty restrictive.  Do we know the context on why this restriction exists?  Not really for PyTorch, but in general adding new fields in future versions is a pretty common thing to do when versioning things.

jerryzh168 replied:
why do we need forward compatibility? 

and why double? is there plans to add double in next few PRs?

vkuzo commented on aten/src/ATen/native/quantized/cpu/fbgemm_utils.cpp:
```diff
@@ -212,99 +213,17 @@ Tensor ConvertToChannelsLast3dTensor(const Tensor& src) {
+          return serialize_conv<kSpatialDim>(params);
         },
         [](SerializationType state)
         -> c10::intrusive_ptr<ConvPackedParamsBase<kSpatialDim>> { // __setstate__
```
should this function be the place where BC is guaranteed for non-ONXX?  I.e. it would know how to read all past supported versions of the data from disk and convert them to the current version.  Or, does that happen elsewhere?

z-a-f replied:
THat's the part that I cannot figure out how to implement

dreiss commented on aten/src/ATen/native/quantized/cpu/serialization_versions.h:
```diff
@@ -0,0 +1,164 @@
+ *    1. bias
+ *  5. list of all doubles
+ *    EMPTY
+ *  6. list of all integers (vector<int64_t> of size (1 + 4 * kSpatialDim + 1))
```
I count 3 * kSpatialDim

z-a-f replied:
Oh this bled through from later PRs -- it is supposed to be 1 + 3 * kSpatialDim. In the next PR it is changed to 1 + 4 kSpatialDim + 1 => groups (1) + strides/padding/dilation/output_padding(kSpatialDim) + transpose
